pin template with documentation, sample "file" pin

cbiedl opened the pull request:
At least for me, writings pins was hard since I had to figure out the concept behind them, and when studying the existing ones, finding out what's generic pin logic and what it specific for that pin wasn't that easy either.

That should not be repeated for anyone else, so I created `src/pins/template/` and filled it my several files and a lot of explanations embedded.

Also a new pin "file" as a demonstration of a minimal pin that uses a file to store the key. I'd consider this suitable for educational puposes only, but you think it's worth to include it in the official distribution, go ahead. Actually, this was im preparation of #185 which will get a fix soon.

Commit:
Provide a template for new pins

This should ease writing new pins a lot, for newcomers and for
experienced authors as well.

sarroutbi commented on src/pins/file/clevis-encrypt-file.1.adoc:
```diff
@@ -0,0 +1,44 @@
+
+== BUGS
+
+Requires the directories for that file already exists.
```
I would rewrite this as:
"Requires that directories for that file already exist"

sarroutbi reviewed (commented):
Minor change in documentation suggestion. Rest of changes LGTM.

cbiedl commented:
Certainly. Not being a native English speaker, such glitches do happen.
Do you want an updated pull request? Might be less work if you just do it on your own.

Commit:
Implement a pin to store the key in a single file

This is a very simple pin, and possibly rather for educational
purposes: Store the jwk at a given place in the file system.

cbiedl commented:
Okay, found some minor issues so I've updated my branch (also, rebased it on top of HEAD). This resulted in some warnings/errors from CI I fail to understand. Feel free to take it as-is.

sarroutbi commented:
> Okay, found some minor issues so I've updated my branch (also, rebased it on top of HEAD). This resulted in some warnings/errors from CI I fail to understand. Feel free to take it as-is.

Thanks for the PR. The differential shellcheck issue looks like a false positive, so no need to worry for that.

zi4k4j74yn commented:
Is there any way one can assist to getting this merged?

sarroutbi commented:
Changes LGTM. @sergio-correia: can you please provide feedback when possible?

sarroutbi reviewed (approved):
Change LGTM. @sergio-correia : can you please provide your feedback when possible?

sarroutbi commented:
@sergio-correia : can you please provide your feedback when possible? I would prefer a second opinion. For me, this LGTM

sergio-correia reviewed (approved):
This looks good to me as well. Thanks, @cbiedl.

oldium commented:
FYI this will not be easy to package for distributors.

For Fedora-based distros this is straight-forward, because it uses just Dracut, so installing `clevis-decrypt-file` in `module-setup.sh` unconditionally is fine.

For Debian-based distros this will not be possible, because you can have either initramfs-tools or Dracut. The support for `initramfs-tools` is covered by `clevis-initramfs` package, support for Dracut is handled by clevis-dracut package.

To make this working in Debian, the hook scripts need to check if the clevis pin files exist prior to installing the full support.

For Dracut `module-setup.sh`, something like this should do the trick (both for `template` and for `file` pins):

`module-setup.sh.in`:
```
check() {
    require_binaries clevis-decrypt-@pin@ || return 1
}
```

For initramfs-tools, something like this should make it working:

```
if [ -x @bindir@/clevis-decrypt-@pin@ ]; then
    copy_exec @bindir@/clevis-decrypt-@pin@ || die 1 "@bindir@/clevis-decrypt-@pin@ not found"
fi
```
